src/utils/Simplify.h

```diff
@@ -0,0 +1,465 @@
+     * If removing a vertex causes the covered area of the line segments to
+     * change by more than this, it may not be removed.
+     */
+    coord_t max_area_deviation;
+
+protected:
+    /*!
+     * Line segments smaller than this should not occur in the output.
+     * If a vertex causes deviation of less than this, it should always be
+     * removed.
+     */
+    constexpr static coord_t min_resolution = 5; //5 units, regardless of how big those are, to allow for rounding errors.
+
+    /*!
+     * The main simplification algorithm starts here.
+     * \tparam Polygonal A polygonal object, which is a list of vertices.
+     * \param polygon The polygonal chain to simplify.
+     * \param is_closed Whether this is a closed polygon or an open polyline.
+     * \return A simplified polygonal chain.
+     */
+    template<typename Polygonal>
+    Polygonal simplify(const Polygonal& polygon, const bool is_closed) const
+    {
+        const size_t min_size = is_closed ? 3 : 2;
+        if(polygon.size() < min_size) //For polygon, 2 or fewer vertices is degenerate. Delete it. For polyline, 1 vertex is degenerate.
+        {
+            return createEmpty(polygon);
+        }
+        if(polygon.size() == min_size) //For polygon, don't reduce below 3. For polyline, not below 2.
+        {
+            return polygon;
+        }
+
+        std::vector<bool> to_delete(polygon.size(), false);
+        auto comparator = [](const std::pair<size_t, coord_t>& vertex_a, const std::pair<size_t, coord_t>& vertex_b)
+        {
+            return vertex_a.second > vertex_b.second || (vertex_a.second == vertex_b.second && vertex_a.first > vertex_b.first);
+        };
+        std::priority_queue<std::pair<size_t, coord_t>, std::vector<std::pair<size_t, coord_t>>, decltype(comparator)> by_importance(comparator);
+
+        //Add the initial points.
+        for(size_t i = 0; i < polygon.size(); ++i)
+        {
+            const coord_t vertex_importance = importance(polygon, to_delete, i, is_closed);
+            by_importance.emplace(i, vertex_importance);
+        }
+
+        //Iteratively remove the least important point until a threshold.
+        Polygonal result = polygon; //Make a copy so that we can also shift vertices.
+        coord_t vertex_importance = 0;
+        while(by_importance.size() > min_size)
+        {
+            std::pair<size_t, coord_t> vertex = by_importance.top();
+            by_importance.pop();
+            //The importance may have changed since this vertex was inserted. Re-compute it now.
+            //If it doesn't change, it's safe to process.
+            vertex_importance = importance(result, to_delete, vertex.first, is_closed);
+            if(vertex_importance != vertex.second)
+            {
+                by_importance.emplace(vertex.first, vertex_importance); //Re-insert with updated importance.
+                continue;
+            }
+
+            if(vertex_importance <= max_deviation * max_deviation)
```

Comment:
I'd prefer not to, for two reasons:
* In this particular case, the compiler will optimise it. While the C++ language doesn't guarantee that, I know that GCC will look for values in loops that can be pre-computed (and most likely Clang has something similar). And even if they didn't, CPU pipelining would likely make this operation free anyway. This is not an argument for pre-computing the `max_deviation` and other properties in the constructor (which the compiler will not optimise). However...
* The automated tests make use of the `max_deviation`, `max_resolution` and `max_area_deviation` properties here to validate the tests. They are public properties at the moment. We could change the fields to be squared, but to prevent confusion I'd also make them protected then. We'd have to sqrt there, or we'd have to change the tests to give them the values of the fields there. Neither of them looks nice.